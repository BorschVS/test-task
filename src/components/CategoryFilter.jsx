import {
  RadioBox,
  RadioButton,
  RadioLabel,
  RadioText,
} from '../styled/CategoryFilter.styled';

export const CategoryFilter = () => (
  <RadioBox component="div">
    <RadioLabel>
      <RadioButton type="radio" name="filter" defaultChecked />
      <RadioText>Самый дешевый</RadioText>
    </RadioLabel>
    <RadioLabel>
      <RadioButton type="radio" name="filter" />
      <RadioText>Самый быстрый</RadioText>
    </RadioLabel>
  </RadioBox>
);

export default CategoryFilter;
