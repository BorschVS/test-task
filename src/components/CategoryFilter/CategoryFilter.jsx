import { useTheme } from "@emotion/react";
import {
  RadioBox,
  RadioButton,
  RadioLabel,
  RadioText,
} from "./CategoryFilter.styled";

const CategoryFilter = () => {
  return (
    <RadioBox component="div">
      <RadioLabel>
        <RadioButton type="radio" name="filter" checked />
        <RadioText>Самый дешевый</RadioText>
      </RadioLabel>
      <RadioLabel>
        <RadioButton type="radio" name="filter" />
        <RadioText>Самый быстрый</RadioText>
      </RadioLabel>
    </RadioBox>
  );
};

export default CategoryFilter;
