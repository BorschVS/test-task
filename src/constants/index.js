export const CHEAP_VALUE = 'cheap';
export const FAST_VALUE = 'fast';
