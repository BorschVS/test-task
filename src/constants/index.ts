export const CHEAP_VALUE: string = 'cheap';
export const FAST_VALUE: string = 'fast';
