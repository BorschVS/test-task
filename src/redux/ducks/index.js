export { default as flights } from './flights';
export { default as form } from './form';
export { default as modal } from './modal';
export { default as images } from './images';