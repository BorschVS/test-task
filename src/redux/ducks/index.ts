export { flightsReducer } from './flights';
export { formReducer } from './form';
export { modalReducer } from './modal';
export { imagesReducer } from './images';
