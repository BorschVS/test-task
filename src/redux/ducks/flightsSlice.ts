import { createSlice, PayloadAction } from '@reduxjs/toolkit';

import { FlightData, FlightsState } from 'types/interfaces';
import { FlightCategory } from 'types/types';

import { FAST_VALUE, CHEAP_VALUE } from '../../constants';

export const GET_FLIGHTS = 'GET_FLIGHTS';
export const SET_FLIGHTS = 'SET_FLIGHTS';
export const SET_FLIGHTS_LOADING = 'SET_FLIGHTS_LOADING';
export const SET_FLIGHTS_ERROR = 'SET_FLIGHTS_ERROR';
export const SET_STOPS_FILTER = 'SET_STOPS_FILTER';
export const SET_ALL_STOPS = 'SET_ALL_STOPS';
export const SET_STOPS_FILTER_STATUS = 'SET_STOPS_FILTER_STATUS';
export const SET_CATEGORY_FILTER = 'SET_CATEGORY_FILTER';

const initialState: FlightsState = {
  availableFlights: [],
  filteredFlights: [],
  stopsFilter: [],
  stopsFilterStatus: false,
  flightCategoryFilter: CHEAP_VALUE,
  loading: false,
  error: null,
};

const flightsSlice = createSlice({
  name: 'flights',
  initialState,
  reducers: {
    setFlights(state, action: PayloadAction<FlightData[]>) {
      state.availableFlights = action.payload
        .slice()
        .sort((prev, next) => prev.price - next.price);
    },
    setFlightsLoading(state, action: PayloadAction<boolean>) {
      state.loading = action.payload;
    },
    setFlightsError(state, action: PayloadAction<Error | string>) {
      state.error = action.payload;
      state.loading = false;
    },
    setCategoryFilter(state, action: PayloadAction<FlightCategory>) {
      if (action.payload === CHEAP_VALUE) {
        state.flightCategoryFilter = action.payload;
        state.filteredFlights = state.filteredFlights
          .slice()
          .sort((prev, next) => prev.price - next.price);
        state.availableFlights = state.availableFlights
          .slice()
          .sort((prev, next) => prev.price - next.price);
      }
      if (action.payload === FAST_VALUE) {
        state.flightCategoryFilter = action.payload;
        state.filteredFlights = state.filteredFlights
          .slice()
          .sort(
            (prev, next) =>
              prev.segments[0].duration - next.segments[0].duration
          );
        state.availableFlights = state.availableFlights
          .slice()
          .sort(
            (prev, next) =>
              prev.segments[0].duration - next.segments[0].duration
          );
      }
    },
    setStopsFilter(state, action: PayloadAction<number[]>) {
      state.stopsFilter = action.payload;
      state.filteredFlights = state.availableFlights.filter((flight) =>
        flight.segments.some((segment) =>
          action.payload.includes(segment.stops.length)
        )
      );
      state.stopsFilterStatus = false;
    },
    setAllStops(state) {
      state.stopsFilter = [0, 1, 2, 3];
      state.filteredFlights = state.availableFlights;
      state.stopsFilterStatus = true;
    },
    setStopsFilterStatus(state) {
      state.stopsFilterStatus = state.stopsFilter.length > 0;
    },
  },
});

export const {
  setFlights,
  setFlightsLoading,
  setFlightsError,
  setCategoryFilter,
  setStopsFilter,
  setAllStops,
  setStopsFilterStatus,
} = flightsSlice.actions;

export default flightsSlice.reducer;
