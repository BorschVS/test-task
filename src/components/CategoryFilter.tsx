import { FC, SyntheticEvent } from 'react';
import { useDispatch } from 'react-redux';

import { CHEAP_VALUE, FAST_VALUE } from '../constants';

import { setCategoryFilter } from '../redux/ducks/flightsSlice';
import { AppDispatch } from 'redux/configureStore';

import {
  RadioBox,
  RadioButton,
  RadioLabel,
  RadioText,
} from 'styled/CategoryFilter.styled';

export const CategoryFilter: FC = () => {
  const dispatch: AppDispatch = useDispatch();

  const handleRadioCheck = (event: SyntheticEvent<HTMLInputElement>) => {
    dispatch(setCategoryFilter(event.currentTarget.value));
  };
  return (
    <RadioBox data-testid="category-filter">
      <RadioLabel>
        <RadioButton
          type="radio"
          name="filter"
          value={CHEAP_VALUE}
          onChange={handleRadioCheck}
          defaultChecked
        />
        <RadioText>Самый дешевый</RadioText>
      </RadioLabel>
      <RadioLabel>
        <RadioButton
          type="radio"
          name="filter"
          value={FAST_VALUE}
          onChange={handleRadioCheck}
        />
        <RadioText>Самый быстрый</RadioText>
      </RadioLabel>
    </RadioBox>
  );
};

export default CategoryFilter;
