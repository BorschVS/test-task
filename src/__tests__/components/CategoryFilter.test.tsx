import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom';

import * as actions from '../../redux/ducks/flightsSlice';

import CategoryFilter from 'components/CategoryFilter';

import { CHEAP_VALUE, FAST_VALUE } from '../../constants';

const mockDispatch = jest.fn();

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useDispatch: () => mockDispatch,
}));

describe('CategoryFilter', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  test('Should click and call dispatch with correct action', async () => {
    const mockSetCategoryFilter = jest.spyOn(actions, 'setCategoryFilter');

    render(<CategoryFilter />);

    const radioButtonCheap = screen.getByTestId('radio-cheap');
    const radioButtonFast = screen.getByTestId('radio-fast');

    expect(radioButtonCheap).toBeInTheDocument();
    expect(radioButtonFast).toBeInTheDocument();

    userEvent.click(radioButtonFast);
    expect(mockDispatch).toHaveBeenCalled();
    expect(mockSetCategoryFilter).toHaveBeenCalledWith(FAST_VALUE);

    userEvent.click(radioButtonCheap);
    expect(mockDispatch).toHaveBeenCalled();
    expect(mockSetCategoryFilter).toHaveBeenCalledWith(CHEAP_VALUE);
  });

  test('CategoryFilter snapshot', () => {
    const { asFragment } = render(<CategoryFilter />);

    expect(asFragment()).toMatchSnapshot();
  });
});
