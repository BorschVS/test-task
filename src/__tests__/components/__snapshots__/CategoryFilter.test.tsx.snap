// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CategoryFilter CategoryFilter snapshot 1`] = `
<DocumentFragment>
  <div
    class="css-mmmkij"
    data-testid="category-filter"
  >
    <label
      class="css-we1pt6"
    >
      <input
        checked=""
        class="css-13zcrbs"
        data-testid="radio-cheap"
        name="filter"
        type="radio"
        value="CHEAP_VALUE"
      />
      <span
        class="css-opnvry"
      >
        Самый дешевый
      </span>
    </label>
    <label
      class="css-we1pt6"
    >
      <input
        class="css-13zcrbs"
        data-testid="radio-fast"
        name="filter"
        type="radio"
        value="FAST_VALUE"
      />
      <span
        class="css-opnvry"
      >
        Самый быстрый
      </span>
    </label>
  </div>
</DocumentFragment>
`;
